website/blog/2025-10-08-multiround-local-merge.mdx

```diff
@@ -0,0 +1,89 @@
+---
+slug: multi-round-local-merge
+title: "Multi-Round Lazy Start Merge"
+authors: [duanmeng, xiaoxmeng, pedroerp]
+tags: [tech-blog, spill, operator]
+---
+
+## Background
+
+Meta introduces Generative Recommenders (GRs) as a new paradigm replacing DLRMs in its paper,
+*[Actions Speak Louder than Words: Trillion-Parameter Sequential Transducers for Generative Recommendations](https://arxiv.org/pdf/2402.17152)*.
+Companies such as ByteDance and Kuaishou have also begun exploring the GR paradigm. A key requirement
+is to merge training data across partitions, say merging hourly partitions into daily ones, while ensuring
+that all rows sharing the same primary key are stored consecutively. This is a common need at both
+Meta and ByteDance.
+
+
+Training data is typically partitioned and bucketed by primary key, with rows sharing the same key
+stored consecutively, so merging across partitions essentially becomes a [multi-way merge problem](https://en.wikipedia.org/wiki/K-way_merge_algorithm#).
+
+Normally, [Apache Spark](https://spark.apache.org/) can be used for this sort-merge requirement — for example, via `CLUSTER BY`.
```

Comment:
maybe explicitly state that this will generate in Spark. Does this generate a sorted shuffle by the key?

Reply:
The shuffle can be eliminated by Spark’s `EnsureRequirements` rule if the data is already bucketed by the key. However, sorting cannot be eliminated, as rows still need to be ordered by the key to ensure they are clustered within each partition.

Reply:
Although we applied bucketing and ordering by key when preparing the training data in production, Spark can eliminate the shuffle when merging training data from multiple hourly partitions. However, each Spark task can only read the files planned from multiple partitions within a split one by one, placing them into the sorter and spilling as needed. Only after all files have been read does Spark perform a sort-merge of the spilled files.